crates/sui-framework/src/natives/crypto.rs

```diff
@@ -41,15 +41,57 @@ pub fn ecrecover(
     let signature = pop_arg!(args, Vec<u8>);
     // TODO: implement native gas cost estimation https://github.com/MystenLabs/sui/issues/3593
     let cost = legacy_empty_cost();
+    match recover_pubkey(signature, hashed_msg) {
+        Ok(pubkey) => Ok(NativeResult::ok(
+            cost,
+            smallvec![Value::vector_u8(pubkey.as_bytes().to_vec())],
+        )),
+        Err(SuiError::InvalidSignature { error: _ }) => {
+            Ok(NativeResult::err(cost, INVALID_SIGNATURE))
+        }
+        Err(_) => Ok(NativeResult::err(cost, FAIL_TO_RECOVER_PUBKEY)),
+    }
+}
+
+fn recover_pubkey(signature: Vec<u8>, hashed_msg: Vec<u8>) -> Result<Secp256k1PublicKey, SuiError> {
     match <Secp256k1Signature as ToFromBytes>::from_bytes(&signature) {
         Ok(signature) => match signature.recover(&hashed_msg) {
-            Ok(pubkey) => Ok(NativeResult::ok(
-                cost,
-                smallvec![Value::vector_u8(pubkey.as_bytes().to_vec())],
-            )),
-            Err(_) => Ok(NativeResult::err(cost, FAIL_TO_RECOVER_PUBKEY)),
+            Ok(pubkey) => Ok(pubkey),
+            Err(e) => Err(SuiError::KeyConversionError(e.to_string())),
         },
-        Err(_) => Ok(NativeResult::err(cost, INVALID_SIGNATURE)),
+        Err(e) => Err(SuiError::InvalidSignature {
+            error: e.to_string(),
+        }),
+    }
+}
+
+/// Ecrecover to eth address directly.
+/// Native implemention of ecrecover in public Move API, see crypto.move for specifications.
+pub fn ecrecover_eth_address(
```

Comment:
What is the main reason of having the Eth related function as part of our native functions, vs an example contract on how to do with the already existing api?

Reply:
to compare against any eth onchain data where the pubkey is not immediately available, recovering directly to address is very helpful. i noticed few eth libraries indeed recover to address https://web3js.readthedocs.io/en/v1.7.5/web3-eth-accounts.html#recovertransaction 

an eth address is derived from a uncompressed pubkey whereas in sui, the pubkey is compressed. accessing the underlying library to decompress is necessary in rust instead of in smart contract

Reply:
I was thinking that uncompressing pubKey might be the function we need as native (as it can be reused in other protocols too), and then use this in Move to implement the eth verification.

Reply:
that works too. i went with ease of implementation. slicing arrays in move and move debug tool was much harder than rust 😕 i can make it work

Reply:
btw there is already a example smart contract to do this `ecdsa.move`

Reply:
agree on that, our goal is to have a rich move api to support applications like this (and other that might need it). Ideally things that will stay as standards, and supporting a native api for Ethereum in particular might not be a good idea as it's updating/evolving, it's not a primitive etc (it if it evolves our api will be deprecated).